### ISSUE TITLE
Blazor: Provide common Promise<void> for when Blazor starts

### ISSUE THREAD

**Issue by mrpmorris** [aspnet/AspNetCore#17504]
### Is your feature request related to a problem? Please describe.
We can currently set autostart="false" on the the Blazor `<script>` tag and then use the JavaScript `Blazor.start({}).then(.........)` to execute JavaScript we need to run as soon as Blazor starts.

Using this approach is not suitable for library developers, because Blazor.start can only be called once.

### Describe the solution you'd like
Please provide a new `Blazor.onStarted` function that accepts a callback to execute once Blazor has started. It should be possible to call this function more than once so many scripts can be notified. If Blazor has already started, the callback should be executed immediately.

**Comment by javiercn** [comment]
@mrpmorris thanks for contacting us.

We'd like to understand more about your scenario. Could you please elaborate on the scenario and the problem you're trying to solve?

**Comment by mrpmorris** [comment]
> @mrpmorris thanks for contacting us.
> 
> We'd like to understand more about your scenario. Could you please elaborate on the scenario and the problem you're trying to solve?

Hi

A simple example is some JavaScript that invokes a static .NET method in order to get an API key that is stored in appsettings.json.

The use case isn't really that important, the problem is that any JavaScript that needs to call a static .NET class on startup cannot do so because there is no way to know when Blazor is ready to service the call.

The app developer can do this by calling `Blazor.start({}).then(.....(` but that's a one-off call. Any libraries that have JavaScript that needs to execute automatically don't have a hook point.

There are work-arounds such as getting some Blazor code to initialize the JavaScript, but that's not really the point. This is about JavaScript invoking Blazor static methods.

The three solutions to this are

1. Use setTimeout - but then you have to set the timeout too high.
2. Use setTimeout in a loop and swallow exceptions - this is quite an aggressive way to achieve the goal.
3. Use Blazor.start() - but that can only be hooked once.
4. Get Blazor to call the JavaScript - that works, but not as neat for something like Google Analytics where you just want the user to add a NuGet reference to your library, add an ApiKey to the settings, and reference a JavaScript.

https://blazor-university.com/javascript-interop/calling-dotnet-from-javascript/calling-static-dotnet-methods/

**Comment by javiercn** [comment]
@mrpmorris I wouldn't recommend any javascript library to automatically perform any action (specially involving dotnet) immediately after being loaded by the browser.

The right pattern to follow here is for the library to expose an init/initialize method or similar and call that from the resulting promise from Blazor.start. So equivalent to what you suggested above

```javascript
await Blazor.start({...});
library.Initialize();
```

**Comment by mrpmorris** [comment]
> @mrpmorris I wouldn't recommend any javascript library to automatically perform any action (specially involving dotnet) immediately after being loaded by the browser.
> 
> The right pattern to follow here is for the library to expose an init/initialize method or similar and call that from the resulting promise from Blazor.start. So equivalent to what you suggested above
> 
> ```js
> await Blazor.start({...});
> library.Initialize();
> ```

I agree that is currently the correct approach but, as you know, library creators like to go for as little boilerplate as possible. 

Compare the instructions
(My suggestion)
1. Add NuGet reference to the package
2. Add a `<script>` reference (would be nice not to have to do this in the future)

(Current process)
1. Add NuGet reference to the package
2. Add the `<script>` tag
3. Alter the _Host.cshtml file and set autostart='false' on the Blazor script
4. Add a new script for your app (or `<script>` section in the _Host.cshtml file)
5. Add a new await call to Blazor.start() (passing in the default values, whatever they are)
6. Then call `library.initialize()` for every library you use

Not even sure if 5+6 would have to be an async function which you need to call, or if the main script can await - but anyway, it's just far too much to get a library consumer to do.

It would be a far superior experience for Blazor developers if the libraries they consume could each just do this

Blazor.addStartedCallback(function() {
  initializeMyLibrary();
});

**Comment by javiercn** [comment]
We've moved this issue to the Backlog milestone. This means that it is not going to happen for the coming release. We will reassess the backlog following the current release and consider this item at that time. However, keep in mind that there are many other high priority features with which it will be competing for resources.

**Comment by mrpmorris** [comment]
@javiercn I think your bot is playing up.

**Comment by javiercn** [comment]
@anurse can you take a look? The bot is doing bot things 😵

**Comment by anurse** [comment]
This is the bug I mentioned in Teams earlier that is fixed. The `Needs: Author Feedback` label was applied but not properly removed when @mrpmorris responded. It's all fixed now.
